pymc_extras/statespace/models/structural/components/regression_dataclass.py

```diff
@@ -0,0 +1,539 @@
+from dataclasses import dataclass, field
+
+import numpy as np
+
+from pytensor import tensor as pt
+
+from pymc_extras.statespace.models.structural.core import Component
+from pymc_extras.statespace.utils.constants import TIME_DIM
+
+
+@dataclass
+class ParameterProperty:
+    name: str
+    shape: tuple[int, ...]
+    dims: tuple[str, ...]
+    constraints: str | None = None
+
+    def __str__(self):
+        base = f"name: {self.name}\nshape: {self.shape}\ndims: {self.dims}"
+        if self.constraints:
+            return base + f"\nconstraints: {self.constraints}"
+        return base
+
+
+@dataclass
+class ParameterProperties:
+    parameters: list[ParameterProperty]
+
+    def get_parameter(self, name: str) -> ParameterProperty | None:
+        return next((p for p in self.parameters if p.name == name), None)
+
+    def __getitem__(self, name: str) -> ParameterProperty:
+        result = next((p for p in self.parameters if p.name == name), None)
+        if result is None:
+            raise KeyError(f"No parameter named '{name}'")
+        return result
+
+    def __contains__(self, name: str) -> bool:
+        return any(p.name == name for p in self.parameters)
+
+    def __str__(self):
+        base = f"parameters: {[parameter.name for parameter in self.parameters]}"
+        return base
+
+
+@dataclass
+class DataProperty:
+    name: str
+    shape: tuple[int, ...]
+    dims: tuple[str, ...]
+    is_exogenous: bool
+
+    def __str__(self):
+        base = f"name: {self.name}\nshape: {self.shape}\ndims: {self.dims}\nis_exogenous: {self.is_exogenous}"
+        return base
+
+
+@dataclass
+class DataProperties:
+    data: list[DataProperty]
+    needs_exogenous_data: bool = field(default=False, init=False)
+
+    def __post_init__(self):
+        for d in self.data:
+            if d.is_exogenous:
+                self.needs_exogenous_data = True
```

Comment:
```suggestion
    @property
    def needs_exogenous_data(self) -> bool:
        return any(d.is_exogenous for d in self.items)
```

Just compute this on demand with a property instead of going through post init hoops. In this case you can also set `frozen=True`.